src/debugger/Debugger.js

```diff
@@ -71,8 +75,8 @@ export class DebugServer {
   // Checking if the debugger needs to take any action on reaching this ast node
   checkForActions(ast: BabelNode) {
     this.checkForBreakpoint(ast);
-
-    // last step: set the current location as the previously executed line
+    this.checkStepIn(ast);
```

Comment:
Request change: This kind of sequencial checking has a problem:
Assume you are at line 10, and you set a breakpoint at line 11. Then you issue a step-into which will land at line11 later. In current design, debugger will stop twice. Once inside checkForBreakpoint(), when user resumes from breakpoint, checkStepIn() will stop it again which is wrong.

Reply:
This problem should happen in the current code. But strangely, I'm not able to reproduce it. Will look into it more.

Reply:
There is a check in _isValidSteppingLocation for whether the current location is the same as the previously stopped location. If so, it will not stop again.

Reply:
It works in this scenario because you are missing one invariant here: checking "this. _stepInData is undefined" in _processStepIn() which means there should only be one step-in request on the flight. Currently, you are silently overwriting the this._stepInData for each step-in instead of explicitly cancel the previous stepper. After adding this invariant, you will see the failure. 
Another way to demonstrate the failure:
1. Current line is 10, while set breakpoint at line 11
2. Step-in from line10 => line11
3. BreakpointManager stops at line11 while step-into stepper is not done yet.
4. You do a "step-over"(note: note step-into) which will not overwrite this._stepInData. 
5. The code comes back and checkStepInto() which will stop again at line11.

The root cause of the problem is that it sequentially notify each managers. Instead, you want all the execution control related managers to be notified and then consolidate a final stop disposition decision.

In summary, you want:
1. Separate stepping logic into its own stepper.
2. At each line, send the current AST to all managers
3. Collect stop decision from each manager and join them into a single decision.